### ISSUE TITLE
.NET 3.5/2.0 core package in Server 2016/2019 - Replacement of old components needed

### ISSUE THREAD

**Issue by PrismaComputer** [hmailserver/hmailserver#302]
There is no .net 3.5/2.0 core package to install in Server 2016 and 2019 (GUI and Core). So hMailserver setup fails.
It's a "feature on demand" you need to install from DVD/ISO with DISM or Powershell. The setup should handle this in any way, at least with an information that or how to do it.

Knowing this it's expectable that Microsoft will remove 3.5/2.0 core package completely in one of the next versions.
So, letting x86 die for better support and easier and more modern development is one thing. But replacing old .net components with new ones should be the same step.

**Comment by RvdHout** [comment]
You can simply add .Net 3.5 (2.0) support using add/remove server roles in Server Manager in both server 2016 and server 2019. I can understand this is somewhat tricky in Windows server Core version and you have to do this using Dism or Powershell, but for GUI version this is a non-issue.




Outlook voor Android downloaden<https://aka.ms/ghei36>

________________________________

**Comment by PrismaComputer** [comment]
1. Martin build in some code to install the package on servers automatically. This code isn't working anymore. So this is an issue.

2. From my knowledge installation of .NET 2.0 with GUI servers using sever manager also fails in first attempt with cryptic error code. After choosing a source path on DVD it's then working, yes...

3. Obviously Microsoft is willing to push out old applications in future. So it's time to act now and not when it's to late.

**Comment by RvdHout** [comment]
1. It just checked if .net 2.x framework is installed on target PC. I agree this ain't working on new windows version but this isn't a server specific check

2. Never had that issue...but on the other hand, what is the problem using dvd/USB with windows source?

3. Don't get your Hopes up

Outlook voor Android downloaden<https://aka.ms/ghei36>

________________________________

**Comment by lubber-de** [comment]
Do this before installing hmaillserver. Works fine on 2016/2019 (GUI)
```
dism.exe /online /enable-feature /featurename:NetFX3 /all
```

**Comment by martinknafve** [comment]
@RvdHout, do you see any issues with switching over to Visual Studio 2019 in master branch? I've made the upgrade locally (but not yet committed anything to git). I'm thinking it's probably time to upgrade to the latest version of Visual Studio and .NET Framework. 

I'm using Visual Studio 2019 Community edition + manual installation of Windows SDK + Windows WDM to get ATL-support. Those are probably built-in into Professional edition but was thinking it's better to stick to a version eveyone can use for free.

**Comment by RvdHout** [comment]
> @RvdHout, do you see any issues with switching over to Visual Studio 2019 in master branch? I've made the upgrade locally (but not yet committed anything to git). I'm thinking it's probably time to upgrade to the latest version of Visual Studio and .NET Framework.
> 
> I'm using Visual Studio 2019 Community edition + manual installation of Windows SDK + Windows WDM to get ATL-support. Those are probably built-in into Professional edition but was thinking it's better to stick to a version eveyone can use for free.

No, that's not a problem, what Windows SDK + Windows WDM to get ATL-support yu are using...or will this be listed in the build instructions once updated?

**Comment by martinknafve** [comment]
I have not pushed anything yet so so far you don't have to do anything. 

Turns out WDM/SDK isn't required to be separately installed (I think). To make it work locally, I've installed  Visual Studio 2019 Community edition. During installation, I've picked:

* C++ ATL for latest v142 build tools (x86 & x64)
* Windows SDK 10.0.18362.0

I had to do a couple of changes to the source as well to get it to compile with the latest compiler (some due to C++ language changes). I had to recompile both Boost and OpenSSL using the latest compiler, and had to change the PostgreSQL client to be dynamically linked instead of statically.

I will also see if I can update the .NET clients to .NET framework 4.8 or if that will cause some issue.

**Comment by RvdHout** [comment]
.NET framework 4.8? Wouldn't it be better to pick a lower version and not jump straight to the latest available version, say .NET 4.0/4.5? otherwise anyone on older systems (like Windows Server 2012 w/o R2 or even Server 2016) first need to update to latest .NET

**Comment by martinknafve** [comment]
I was just thinking since there's no guarantee that they will have .NET Framework 4.5 installed and may have to install it first, we could just pick the latest version. I tried installing 4.8 on Windows Server 2008 and it takes a few minutes.

.NET Framework 3.5 is included in Windows Server 2008 and v4.5 in Windows Server 2012. So targeting 4.0 probably does not make much sense, right? I suspect 4.5 has a wider usage than 4.0 (4.0 felt like an in-between version to me at least).

But we can go with .NET 4.5.

**Comment by RvdHout** [comment]
Yeah, i found/added the wikipedia page above later, 4.5 seems reasonable

I think if you would use .NET 4.5 nobody on Windows Server 2012 (extended support until 10/10/2023) or later has to install anything specific to be able to run the hMailServer Administrator Utility

**Comment by martinknafve** [comment]
I've updated the instructions on how to build the master branch. If you give it a try then let me know how it goes and if I've missed something @RvdHout.

**Comment by RvdHout** [comment]
On a pretty fresh Server 2019 OS, the latest build 5.7.0-B2511 gives error during install
Server would not run, after installing https://aka.ms/vs/16/release/vc_redist.x64.exe it runs

Why you say 'Start a x86 Native Tools Command Prompt for VS2019' for building boost & openssl?
i used a x64 Native Tools Command Prompt for VS2019. at it builds and runs fine

**Comment by martinknafve** [comment]
"x86 Native Tools" was a typo. Fixed that.

I'll try on Server 2019 myself and will update the installation to include the proper vc++ runtime dll's. The installation still bundle the old ones and I guess my fresh 2008-install just happened to have the proper DLL:s installed already.

**Comment by RvdHout** [comment]
If i am not mistaken, i think it were errors on MSVCP140.dll, VCRUNTIME140.dll (and/or VCRUNTIME140_1.dll)

**Comment by martinknafve** [comment]
I see the installation ships the wrong DLL:s and I could just update it, but I wanted to make sure I could reproduce the issue. So I installed a new Windows Server 2019 Standard Edition in a virtual machine but then it worked fine - my clean installation of 2019 have the DLL:s pre-installed in the System32 directory. (I did do a full run of Windows Update before trying to install it though).

I can see what DLL's hMailServer.exe loads and make sure those are shipped as part of the installation. I'll do some reading on the subject.

**Comment by RvdHout** [comment]
@martinknafve That sounds hard to believe, just did a test myself as well, using en_windows_server_2019_updated_april_2020_x64_dvd_12d6dc63.iso

In order of appearance (this is during installtion)

![Untitled1](https://user-images.githubusercontent.com/13403301/80318250-3005ef80-8809-11ea-945c-4fd464fbd149.png)

![Untitled2](https://user-images.githubusercontent.com/13403301/80318254-33997680-8809-11ea-95fe-8a0c0af9129f.png)

![Untitled3](https://user-images.githubusercontent.com/13403301/80318256-372cfd80-8809-11ea-9fa2-cd42d29bcffb.png)

![Untitled4](https://user-images.githubusercontent.com/13403301/80318258-3a27ee00-8809-11ea-98f4-99ab6a53c510.png)

**Comment by OutbackMatt** [comment]
Just trying to install the 2514 build on my (production) windows 10 system got the exact same errors
I've reverted back to 2507 (which is working great btw)

Also, the 2514 build didn't stop my running hmailserver before installing, I had to manually stop the hmailserver service.

**Comment by martinknafve** [comment]
@RvdHout You're right. I did more some digging, and the clean Windows Server 2019 installation did in fact not include these files. The reason I had them is because I'm using VMware to host my virtual machines, and by default when the VM is created in VMware Workstation then it installs a piece of software called "VMware Workstation Tools" (which provides copy/paste, improved graphics drivers etc for the VM). Turned out that this tools package included the same DLL:s as hMailServer use. I was able to bypass this by creating the VM in a different way where VMWare tools is not installed.

**Comment by RvdHout** [comment]
@martinknafve, now you are updating the development branch, might it be worth updating to the latest Boost as well? (Version 1.72.0)

**Comment by RvdHout** [comment]
@OutbackMatt 
The goal is to remove the dependency for .Net 3.5 (2.0) support (mainly for the Admin GUI) as this isn't installed by default on modern Windows OS's

**Comment by Dravion** [comment]
The following Projects are set to target .NET 2.0 Framework 
hMailAdmin,exe (.NET Executable)
DataDirectorySynchronizer.exe (.NET Executable)
DBSetup.exe (.NET Executable)
DBSetupQuick.exe (.NET Executable)
DBUpdater.exe (.NET Executable)
Shared.dll .NET DLL)
![net-20](https://user-images.githubusercontent.com/15117429/80405521-07423080-88c3-11ea-93b2-6bdc9c23a642.png)

Only hMailServer.exe is a native Windows PE Executable which means it will run on Windows 
without any .NET Framework stuff installed.

However: From Windows 7 upto Win10/2019 (verified) the simple Windows Command prompt
Command will solve all the trouble: 
DISM /Online /Enable-Feature /FeatureName:NetFx3 

This can be integrated in Innosetup like this:
#Innosetup / Prepare phase

[Code]
function inetnetfx3():Boolean;
   var ReturnCode: Integer;
 begin
  Exec('Dism', 'DISM /Online /Enable-Feature /FeatureName:NetFx3', '', SW_SHOW, ewWaitUntilTerminated, ResultCode);

 if ReturnCode = 0 then    
        ReturnCode := True
      else
        MsgBox('ERROR: DISM.EXE Return code says '+SysErrorMessage(ReturnCode), mbCriticalError, MB_OK);
 end;

There is no need to check if it is installed or not. DISM.EXE will allways return with Code 0
if .NETFX3 Components are installed or it has installed. 

You can check it allways like this: echo status is %errorlevel%

**Comment by RvdHout** [comment]
Thanks you for this very, very complete and not to forget unnecessary reply... i am sure both @martinknafve as myself are aware of what is written in C++ and C#

There is simply no longer a reason to keep the C# stuff at .Net 2, everything prior and including Windows 7 is now EOL, so no more automatic support for that OS is needed, and Win 7 user can still install Net 4.5 to make it work

**Comment by Dravion** [comment]
> Thanks you for this very, very complete and not to forget unnecessary reply... i am sure both @martinknafve as myself are aware of what is written in C++ and C#
> 
> There is simply no longer a reason to keep the C# stuff at .Net 2, everything prior and including Windows 7 is now EOL, so no more automatic support for that OS is needed, and Win 7 user can still install Net 4.5 to make it (we are talking about 5.7.x only here) work

Looks like you cannot read, NETFX3 is version 3 of the .NET Framework and the recommended Pascal Script code invokes a Install Command which works on all modern Win10 and Win 2019
and additionally even on Win7.. Try to read and understand before you talk crap.

**Comment by RvdHout** [comment]
bla bla bla....We are talking about build-in support, without the need to install additional packages
Master branch is already switched to .Net 4.5....so i am not sure what you trying to accomplish here?

**Comment by Dravion** [comment]
Welcome on Ignore, idiot

**Comment by RvdHout** [comment]
I am well aware .Net 3.5 (2.0) support can be added, read the first reply in this topic.... now who is the idiot?

**Comment by RvdHout** [comment]
@martinknafve 5.7.0-B2518 installed without issues 👍

**Comment by RvdHout** [comment]
@martinknafve 

Another issue
With the DDL's in the Programs /Bin directory i get a error on the DCOM (webadmin would not load)
After installing vc_redist.x64.exe the error goes away

**Comment by RvdHout** [comment]
Might it be possible i had the singlethe VC++ Redistributable for 2017 installed before and this was overwritten by the Microsoft Visual C++ Redistributable for Visual Studio 2015, 2017 and 2019? 

Otherwise i can't explain what is happening now....

**Comment by OutbackMatt** [comment]
2518 installed for me  fine ALTHOUGH I had to manually stop both the hMailserver service and the hMailAdmin GUI logon screen

The service did stop, but restarted (as per windows service settings), I had to manually stop it again to install the update
The admin Gui did shut, but the logon screen came up before the update had installed

**Comment by PrismaComputer** [comment]
First of all, thank you for taking care of this topic. Replacing old components with native shipped ones makes sense. And I read, nearly everybody sees that.
But please let me annotate one thing: The Image I used for install had FX3 as feature-on-demand which means once had to dism with path to DVD. But hey, no big deal. But this fact made me think that it's time to replace the components. That's all...

Maybe your images were different, but that's not important anymore. Go ahead, and thank you!

**Comment by RvdHout** [comment]
@PrismaComputer 
Be aware, this change only is pushed into the 5.7.x development branch

You can get those builds by going here https://build.hmailserver.com/login.html, then login as guest, select the Build hMailServer 5.7 branch, expand the list, select the topmost and go to artifacts to download

**Comment by RvdHout** [comment]
@martinknafve 
Problem with Server 2012, 5.7.0-B2518 using build-in MSSQLCE DB (DB creation fails)

![svr2012](https://user-images.githubusercontent.com/13403301/80468285-46ae6280-893f-11ea-819d-cd8868857945.PNG)

**Comment by PrismaComputer** [comment]
THX for making that clear, I'm aware of this. 

PS: Just checked dism /online /get-features:
Featurename : NetFx3
Status : Deaktiviert und Nutzlast entfernt

**Comment by PrismaComputer** [comment]
Uuups, why closed?

**Comment by RvdHout** [comment]
> @martinknafve
> Problem with Server 2012 (Hyper-V test), 5.7.0-B2518 using build-in MSSQLCE DB (DB creation fails)
> 
> ![svr2012](https://user-images.githubusercontent.com/13403301/80468285-46ae6280-893f-11ea-819d-cd8868857945.PNG)

Same on Windows Server 2012 R2

**Comment by RvdHout** [comment]
https://www.microsoft.com/en-us/download/details.aspx?id=30709

> .NET Framework 3.5 SP1 or .NET Framework 4 needed for managed application development.

FYI, hashes on above downloads are different than the once packed with hmailserver

**Comment by RvdHout** [comment]
@martinknafve 

LOL... already expected this would happen :)

> April 28th, 2020 03:57 GMT, Boost Version 1.73.0

**Comment by martinknafve** [comment]
@RvdHout Well that's just silly :)

**Comment by RvdHout** [comment]
> > @martinknafve
> > Problem with Server 2012 (Hyper-V test), 5.7.0-B2518 using build-in MSSQLCE DB (DB creation fails)
> > ![svr2012](https://user-images.githubusercontent.com/13403301/80468285-46ae6280-893f-11ea-819d-cd8868857945.PNG)
> 
> Same on Windows Server 2012 R2, but no issue on Server 2016

Error seems not isolated MSSQLCE , when trying to install with MySQL i get the same error (api-ms-win-crt-runtime-l1-1-0.dll)

**Comment by RvdHout** [comment]
On a fully updated (geez that took a while!) Windows Server 2012 Re (Hyper-V) 5.7.0-B2518 installs without this error, so somewhere running Windows Updates it installs the required api-ms-win-crt-runtime-l1-1-0.dll

**Comment by RvdHout** [comment]
On Windows Server 2012 the installer works after installing just KB2999226 (Windows8-RT-KB2999226-x64.msu)
It didn't require any prerequisite KB to be installed first like with Windows Server 2012 R2

**Comment by Dravion** [comment]
Regarding OS Support:
The new, updated build instructions for hMailServer targeting WIN32_WINNT=0x600 (Vista x64 and Windows Server 2008 x64). 

However:  
Your recommended Windows8-RT-KB2999226-x64.msu update requires at least Vista x64 + SP2 and Windows Server 2008 x64 with SP2 installed on top of it. Next we need .NET 4.5 installation via "Dism /online /Enable-Feature /FeatureName:NetFx4 /All".

Maybe its also possible to trigger a Online Windows Update installation of KB2999226 via hMailServer-Setup for Vista SP2 / Windows  Server 2008 SP2 / Win7 SP1 / Server 2008_R2 / Win 8.0+Win 8.1 / Server 2012 and Windows Server 2012_R2 but this has to be done before hMailServer.exe is started the first time (before the DBSetup step). 

It's a lot of checking, but if Martin wishes to stay compatible with Vista / Server 2008 this should be done by the installer somehow IMHO.

**Comment by RvdHout** [comment]
Should we really have to worry for prerequisites on those older systems? One might hope they all updated with latest Windows Updates.  And come on, Vista???  Even Windows 7 is EOL and should no longer be supported out of the box ...why are we talking about Vista?

**Comment by mstenz** [comment]
> why are we talking about Vista?

Because always newest Windows and Updates is only theoretical but not practival. I have had my hmailserver till last month running even on a SRV2018 machine. As long there is no hard requirement (a new feature is only running with the new type of software) i don't see any need for forcing others to update. For some Windows Server licenses costs thousands of dollars and they cannot spend it every 2-3 years.

**Comment by RvdHout** [comment]
@mstenz, Server 2008 including R2 is EOL..... you can still make it work, but is should not be supported out of the box. You still can make it work but you have to meet prerequisites yourself for it to run, like: Net 4.5, KB2999226

**Comment by mstenz** [comment]
> you can still make it work, but is should not be supported out of the box.

But why? Just because on paper from Microsoft it is out of there regular support doesnt mean users have this version of OS in use. And as said. If there is no requirement in hmailserver that forces the use of a newer OS, why we should manually block the installation. We all know the software will run without any issues (because it has in the past), the only thing what would make sense is a warning maybe to the users they user older OS but I would definitive not prevent the installation. But also this (the warning) is i think not the job of hmailservers. Users are responsible for their own system so even this normally is not needed.

**Comment by RvdHout** [comment]
Just because...we can't keep supporting the antiques with all side effects what is required to run it out of the box, you would get a 1TB installer

Note: we are talking 5.7 branch only here, if 5.6.x works for you keep using it....

**Comment by mstenz** [comment]
@RvdHout i am using already 5.7. It works without issues on Server 2018 (if you remove the blocker in installation)

**Comment by RvdHout** [comment]
2008!!!

**Comment by RvdHout** [comment]
@Dravion 
I doubt you can install .NET 4.5 via "Dism /online /Enable-Feature /FeatureName:NetFx4 /All" on Widows Server 2018 and Vista...it simply not part of the installation, see: https://en.wikipedia.org/wiki/.NET_Framework_version_history

**Comment by RvdHout** [comment]
@mstenz 
I would like to compare it with lets says a iPhone 5 or Samsung Galaxy S6.... do they support the latest IOS and Android versions? No! Can it run it? Probably....some people done it by rooting the device, they did their own prerequisites to make it work

**Comment by mstenz** [comment]
I don't really want to fight for this with you. My standpoint here is clear. As long there is no technical change in hmailserver that make it incompatible with any OS or prerequisite that is not able to make - only then we should stop supporting this OS. Otherwise it should still be allowed to use it for all users without any need to reconfigure installers or others for everyone. 
Also i am no friend of including all and everything in an installer (as you say over time it gets bigger). Instead write down an installation instruction and allow downloading of the components from the hmailserver website or link to them on for example microsoft download site. This is much more effective. We ware talking here still  about a SERVER application. It is by nature that people using it should have more knowledge then standard PC "users" in installation practices and how to manage a system. So if the documentation say clear XYZ are the requirement and you can download it here if not yet installed then everything should be fine. I do not think hmailserver installer need to do all automatically because that is only waste of time on our end.

**Comment by RvdHout** [comment]
I am not fighting with you...did i? Anyway, this is still **only my** opinion, after all it is martin who decides.
The point is you will get a unmanageable installer if all prerequisites for every OS before, lets says: Windows Server 2012 (this is the first that included .NET 4.5 out of the box) has to be met/covered

**Comment by RvdHout** [comment]
@Dravion
You seem to know you way around in Innosetup, is it possible to add a check for a KB to be present/installed on the System?

**Comment by RvdHout** [comment]
I think here i have listed the full list of requirements for running hMailServer 5.7.x on Windows Server 2008 and up

**Windows Server 2008** 
(**NOTE:** InnoSetup doesn't allow installation with 5.7.0-B2518, but i think it could run without issues)
Requirements:
- Service Pack 2 (Windows6.0-KB948465-X64.exe)
- Windows6.1-KB2999226-x64.msu
- Net 4.5

**Windows Server 2008 R2**
Requirements:
- Windows6.1-KB2999226-x64.msu
- Net 4.5

**Windows Server 2012**
Requirements:
- Windows8-RT-KB2999226-x64.msu

**Windows Server 2012 R2**
Requirements:
- Windows8.1-KB2919442-x64.msu (required for KB2919355)
- Windows8.1-KB2919355-x64.msu (required for KB2999226)
- Windows8.1-KB2999226-x64.msu

**Windows Server 2016**
Requirements: n/a

**Windows Server 2019**
Requirements: n/a

**Comment by Dravion** [comment]
> I think here i have listed the full list of requirements for running hMailServer 5.7.x on Windows Server 2008 R2 and up

Same goes for Win 7 x64 + SP1 + KB2999226
I was running a series of VM Windows Tests for hMailServer 5.7.0-B2518-18-x64 
![hMailServer-5 7 0-B2518-x64-Vista-SP2_Test-FAIL](https://user-images.githubusercontent.com/15117429/80920000-4feb6500-8d6d-11ea-8e7e-3abab0a88249.png)

![hMailServer-5 7 0-B2518-x64-Win7_SP1_Test-PASS](https://user-images.githubusercontent.com/15117429/80920002-5083fb80-8d6d-11ea-9a54-632efec794eb.png)
![hMailServer-5 7 0-B2518-x64-Win2008_R2_SP1-Test-PASS](https://user-images.githubusercontent.com/15117429/80920004-51b52880-8d6d-11ea-8cc2-98501fe73fb8.png)
![hMailServer-5 7 0-B2518-x64-Win2008_SP2_Test-FAIL](https://user-images.githubusercontent.com/15117429/80920005-51b52880-8d6d-11ea-895a-d280a9726144.png)

**Comment by Dravion** [comment]
> @Dravion
> You seem to know you way around in Innosetup, is it possible to add a check for a specific KB to be present/installed on the System?

Yes, there is a way to connect to the Windows Update Agent from Innosetup via COM Component 
and you can search for a KB Update or using the WMI WQL Query. There is also a way to trigger the installation of the KB-Update. If will look into this if i have a little bit of time.

**Comment by RvdHout** [comment]
> > @Dravion
> > You seem to know you way around in Innosetup, is it possible to add a check for a specific KB to be present/installed on the System?
> 
> Yes, there is a way to connect to the Windows Update Agent from Innosetup via COM Component
> and you can search for a KB Update or using the WMI WQL Query. There is also a way to trigger the installation of the KB-Update. If will look into this if i have a little bit of time.

The way i was thinking is more only checking if a specific KB is installed, and show a dialog, like above if not.... installation makes it unnecessary complicated, especially because it involves ancient OS's like Vista and Windows 7

**Comment by RvdHout** [comment]
https://stackoverflow.com/questions/36358221/how-to-determine-whether-a-specific-windows-update-package-kb-msu-is-installe

First example seems but faster

**Comment by RvdHout** [comment]
```
function IsKBInstalled(KB: string): Boolean;
var
  WbemLocator: Variant;
  WbemServices: Variant;
  WQLQuery: string;
  WbemObjectSet: Variant;
begin
  WbemLocator := CreateOleObject('WbemScripting.SWbemLocator');
  WbemServices := WbemLocator.ConnectServer('', 'root\CIMV2');

  WQLQuery := 'select * from Win32_QuickFixEngineering where HotFixID = ''' + KB + '''';

  WbemObjectSet := WbemServices.ExecQuery(WQLQuery);
  Result := (not VarIsNull(WbemObjectSet)) and (WbemObjectSet.Count > 0);
end;

function IsWindowsVersionOrNewer(Major, Minor: Integer): Boolean;
var
  Version: TWindowsVersion;
begin
  GetWindowsVersionEx(Version);
  Result :=
    (Version.Major > Major) or
    ((Version.Major = Major) and (Version.Minor >= Minor));
end;

function IsWindowsVistaOrNewer: Boolean;
begin
  Result := IsWindowsVersionOrNewer(6, 0);
end;

function IsWindows10OrNewer: Boolean;
begin
  Result := IsWindowsVersionOrNewer(10, 0);
end;

function InitializeSetup(): Boolean;
begin
  if IsWindowsVistaOrNewer and not IsWindows10OrNewer then
    begin
    if IsKBInstalled('KB2999226') then
      begin
        MsgBox('KB2999226 is installed', mbInformation, MB_OK);
      end 
    else
      begin
        MsgBox('KB2999226 is not installed', mbInformation, MB_OK);
    end;
  end;
end;

```

**Comment by Dravion** [comment]
Ok, i tested it
The WQL Query works fine (Windows 7-x64+SP1 / Windows 2008_R2-x64+SP2 and above,
but not Windows Server 2008 and Vista).

![kb_is_not_installed](https://user-images.githubusercontent.com/15117429/80990277-627da100-8e36-11ea-9ad6-c0f60df5a589.png)

and

![kb_is_installed](https://user-images.githubusercontent.com/15117429/80990292-67daeb80-8e36-11ea-9982-7125bd8abbff.png)

I was fiddling arround with the Windows Update Agent, but it looks like the required KB Update we
need isn't available in the Windows Update Catalog, so it needs to be downloaded and installed manually  or downloaded and installed by Innosetup itself. 

[mysetup.zip](https://github.com/hmailserver/hmailserver/files/4575922/mysetup.zip)


Attached ZIP contains the standalone test installer, just testing this script
EXE + ISS Script
[mysetup.zip](https://github.com/hmailserver/hmailserver/files/4575939/mysetup.zip)

**Comment by Dravion** [comment]
Ok, more bad news.
Instead of trying to downloading and installing KB2999226, i was simply trying to unpack the Windows Update package (*.msu) and extract the required files. But to my surprise, any OS 
(which is not Windows 10 or Windows Server 2016 or above) has its own KB2999226 Update Installer.

Update for Windows 7 for x64-based Systems (KB2999226)
https://www.microsoft.com/en-us/download/details.aspx?id=49093

Update for Windows 8 for x64-based Systems (KB2999226)
https://www.microsoft.com/en-us/download/details.aspx?id=49082

Update for Windows 8.1 for x64-based Systems (KB2999226)
https://www.microsoft.com/en-ie/download/details.aspx?id=49081

Update for Windows Server 2008 R2 x64 Edition (KB2999226)
https://www.microsoft.com/en-us/download/details.aspx?id=49062

Update for Windows Server 2012 R2 x64 (KB2999226)
https://www.microsoft.com/en-us/download/details.aspx?id=49063

Maybe we should just check if KB2999226 is installed or not and showing the 
correct download links and quitting the installer until the Update was installed.

**Comment by Dravion** [comment]
Ok, there is a Testinstaller available.
https://github.com/Dravion/hmailserver/releases/tag/5.7.0-B2518-x64-Test

**Comment by RvdHout** [comment]
Personally i would have not chosen to include all them KB2999226 packages in the installer (you might wanna clean this up as well, those files stay behind in /Bin) for that 1% that might try to install hMailServer on older Windows version

I got an error trying to install hmailserver on 2012 R2, during KB2999226 installation, i think i might had to reboot first after installing Windows8.1-KB2919442-x64.msu and Windows8.1-KB2919355-x64.msu

Why not add Server 2008 (SP2 required) support as well? Should run....

**Comment by Dravion** [comment]
Windows Server 2008 is the Server variant of Windows Vista and shares the same internal 
build number (6.0) while Windows Server 2008 R2 on the other side is the Server variant of Windows 7 which has build number (6.1).

However: 
Martin has decided to support only Windows versions equal or higher to Build number 6.1.
In fact, its even a bit higher then this, because Update Windows6.1-KB2999226-x64.msu requires "Windows 7 Service Pack 1" which increments the Windows Build number to 6.1.7601, the same
is true for Windows Server 2008 R2 and thats why Windows6.1-KB2999226-x64.msu  is the same
Update for Win 7 as well as for Windows Server 2008 R2.

**Comment by Dravion** [comment]
> Windows8.1-KB2919355-x64.msu

You better restart Windows Server 2008 R2 after this Update, its 707 MBytes big (packed)
and replaced hugh chunks of the OS.

**Comment by RvdHout** [comment]
> > Windows8.1-KB2919355-x64.msu
> 
> You better restart Windows Server 2008 R2 after this Update, its 707 MBytes big (packed)
> and replaced hugh chunks of the OS.

Do you read anything thats written here? If i do it manually no reboot is required but with your latest installer it fails installing KB2919355, just telling....do with it whatever you want

**Comment by RvdHout** [comment]
> Windows Server 2008 is the Server variant of Windows Vista and shares the same internal
> build number (6.0) while Windows Server 2008 R2 on the other side is the Server variant of Windows 7 which has build number (6.1).
> 
> However:
> Martin has decided to support only Windows versions equal or higher to Build number 6.1.
> In fact, its even a bit higher then this, because Update Windows6.1-KB2999226-x64.msu requires "Windows 7 Service Pack 1" which increments the Windows Build number to 6.1.7601, the same
> is true for Windows Server 2008 R2 and thats why Windows6.1-KB2999226-x64.msu is the same
> Update for Win 7 as well as for Windows Server 2008 R2.

I am well aware of all this, not sure if martin made that choice deliberately as I see no practical limitation to also support Vista/Server 2008 (with SP2, needed for KB2919355) OpenSSL is build with support for 6.0. not?
```
Perl Configure no-asm VC-WIN64A --prefix=%cd%\out64 --openssldir=%cd%\out64 -D_WIN32_WINNT=0x600
nmake clean   
nmake install_sw
```

**Comment by Dravion** [comment]
Did you read anything regarding KB2919355?
Its not part of the Installer, it requires KB2919442 before it can be installed and only if both are installed correctly the Innosetup Installer will try to install KB299922. Just read the change log, its not hard.

**Comment by RvdHout** [comment]
Reading is hard i guess?

First i installed
1. KB2919442 
2. KB2919355

Then i ran the installer, it failed on installing KB2999226

But if i install KB2999226 manualy and then run the installer it runs fine, even without rebooting...

**Comment by Dravion** [comment]
I don't know what you are doing, but my tests showing different results:

1) Clean Install of  the Windows Server 2012 R2 x64
2) Installing KB2919442 and KB2919355
3) Letting the Installer run, if it finds KB2919355 it installs KB299922  

hMailServers is installed and runs as expected.

Works without any Problems on my side.

**Comment by RvdHout** [comment]
Repeated the steps, 

![Untitled](https://user-images.githubusercontent.com/13403301/81440801-e6a29200-9170-11ea-8a3d-d14557320aa1.png)

Installer fails at this point and no database/user is written, it reboots out of is own....or maybe crashed? After reboot KB299922 seems installed though

**Comment by RvdHout** [comment]
Coming back on Vista/Server 2008 support, requirements:
1. Service Pack 2 (Windows6.0-KB948465-X64.exe, required for KB2999226 and Net 4.5)
2. Net 4.5
3. Windows6.0-KB2999226-x64.msu

So in theory it should and might run without issues

**Comment by Dravion** [comment]
> Repeated the steps,
> 
> ![Untitled](https://user-images.githubusercontent.com/13403301/81440801-e6a29200-9170-11ea-8a3d-d14557320aa1.png)
> 
> Installer fails at this point and no database/user is written, it reboots out of is own....or maybe crashed? After reboot KB299922 seems installed though

This message box is raised if the result code from wusa.exe returns with anything else then 0,
which means wusa.exe was not able to install KB2999226 Update.

**Comment by Dravion** [comment]
> Coming back on Vista/Server 2008 support, requirements:
> 
> 1. Service Pack 2 (Windows6.0-KB948465-X64.exe, required for KB2999226 and Net 4.5)
> 2. Net 4.5
> 3. Windows6.0-KB2999226-x64.msu
> 
> So in theory it should and might run without issues

yep

**Comment by RvdHout** [comment]
> Coming back on Vista/Server 2008 support, requirements:
> 
> 1. Service Pack 2 (Windows6.0-KB948465-X64.exe, required for KB2999226 and Net 4.5)
> 2. Net 4.5
> 3. Windows6.0-KB2999226-x64.msu
> 
> So in theory it should and might run without issues

Running with Net 4.5.2 (required for MySQL installer)

![svr2008](https://user-images.githubusercontent.com/13403301/81470801-168f7b00-91ed-11ea-85ad-824f7d7812e3.jpg)

**Comment by martinknafve** [comment]
I'm fine with supporting 2008 (without R2) as well. Only reason I wanted to remove old versions is because the more variants are supported, the more testing is also required.
